### ISSUE TITLE
Constant propagation/folding

### ISSUE THREAD

**Issue by mhyee** [reactorlabs/sourir#73]
I've been pair programming this with @o-.

**Comment by mhyee** [comment]
I've decided to punt on copy propagation. The naive implementation of copy propagation won't work:

Original program:
```
mut x = 0
const y = x
x <- 1
print y   # Outputs 0
```

After constant propagation:
```
mut x = 0
const y = x
x <- 1
print x   # Outputs 1
```

Of course, I could check that `x` is a constant variable before propagating it. Maybe this is simple to implement, but I think focusing on constant folding will give us more interesting examples. I can always revisit this if we decide it's something we want.

**Comment by mhyee** [comment]
Constant folding and constant propagation rely on each other:

```
const x = 1
const y = (2 + x)  # Requires constant propagation of x
const z = (3 + y)  # Requires constant folding of y, then constant propagation of y
```

To fully optimize the above code, you need to perform propagation, then folding, then propagation again, then folding again.

My original plan was to keep the constant propagation and constant folding as small and separate optimizations, and then iterate as long as some optimization made a change. But keeping track of this "changed" status was getting annoying.

I talked to @gasche and he suggested merging constant folding, constant propagation, and copy propagation into a single pass.

The plan now is to maintain an environment (of variables to constants or constant variables) and apply a `normalize` function to expressions. `normalize` would perform any substitutions it could, and then constant fold if possible. If this results in a new constant definition, then that definition would be added to the environment. `drop` instructions would then remove mappings from the environment.

In other words, constant propagation must be done "on the fly," where before, all the constant definitions would be collected upfront, and then all the substitution targets would be computed. This makes sense, since constant folding may introduce new constant definitions.

Unfortunately, this means throwing away all the previous constant folding/propagation code (which was why I avoided this idea in the first place). The code also becomes more complicated, but that seems to be unavoidable.

(I'll probably still leave the dead constant definitions alone. Deleting them messes with the instruction stream and shifts code around.)

**Comment by o-** [comment]
That sounds like an excellent plan! Would the graph representation of the IR help you here? Because I am still thinking about doing this refactoring...

**Comment by mhyee** [comment]
For the actual optimization, I don't think it would help. `successors_at` already gives me what I need. (Maybe once I start writing code, I'll run into a case where the graph representation helps.)

For deleting dead definitions, it would definitely help. But I'd still prefer to delegate this to a later (existing) optimization.

**Comment by mhyee** [comment]
Implemented in #125.
